### ISSUE TITLE
Potential snprintf truncation

### ISSUE THREAD

**Issue by Chocobo1** [arvidn/libtorrent#3719]
**Please provide the following information**

libtorrent version (or branch):
libtorrent 1.2 branch 28d5b7d427df68a5ddfeb29b9c190932ab6fa31a

platform/architecture:
arch linux x64

compiler and compiler version:
gcc version 8.2.1

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.

gcc emits the following warning:
```
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp: In member function ‘void libtorrent::upnp::get_ip_address(libtorrent::upnp::rootdevice&)’:
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:40: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size between 1920 and 1923 [-Wformat-truncation=]
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
                                        ^~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~                     
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~              
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       
   "%s"
   ~~~~                                  
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:1089:10:
  post(d, soap, soap_action);
          ~~~~                           
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:15: note: ‘snprintf’ output 126 or more bytes (assuming 2176) into a destination of size 2048
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
  ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "%s"
   ~~~~         
   , d.path.c_str(), d.hostname.c_str(), d.port
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , int(strlen(soap)), d.service_namespace.c_str(), soap_action
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , soap);
   ~~~~~~~      
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp: In member function ‘void libtorrent::upnp::delete_port_mapping(libtorrent::upnp::rootdevice&, libtorrent::port_mapping_t)’:
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:40: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size between 1923 and 1926 [-Wformat-truncation=]
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
                                        ^~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~                     
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~              
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       
   "%s"
   ~~~~                                  
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:885:10:
  post(d, soap, soap_action);
          ~~~~                           
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:15: note: ‘snprintf’ output 123 or more bytes (assuming 2173) into a destination of size 2048
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
  ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "%s"
   ~~~~         
   , d.path.c_str(), d.hostname.c_str(), d.port
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , int(strlen(soap)), d.service_namespace.c_str(), soap_action
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , soap);
   ~~~~~~~      
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp: In member function ‘void libtorrent::upnp::create_port_mapping(libtorrent::http_connection&, libtorrent::upnp::rootdevice&, libtorrent::port_mapping_t)’:
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:40: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size between 1926 and 1929 [-Wformat-truncation=]
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
                                        ^~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~                     
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~              
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       
   "%s"
   ~~~~                                  
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:766:10:
  post(d, soap, soap_action);
          ~~~~                           
/home/xxx/libtorrent-rasterbar-1_2-git/src/libtorrent/src/upnp.cpp:707:15: note: ‘snprintf’ output 120 or more bytes (assuming 2170) into a destination of size 2048
  std::snprintf(header, sizeof(header), "POST %s HTTP/1.1\r\n"
  ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Host: %s:%d\r\n"
   ~~~~~~~~~~~~~~~~~
   "Content-Type: text/xml; charset=\"utf-8\"\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "Content-Length: %d\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~
   "Soapaction: \"%s#%s\"\r\n\r\n"
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   "%s"
   ~~~~         
   , d.path.c_str(), d.hostname.c_str(), d.port
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , int(strlen(soap)), d.service_namespace.c_str(), soap_action
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   , soap);
   ~~~~~~~      
```

One easy solution is enlarge the `header` array to 3072 or 4096, however I also wonder is it a good idea to have such large variable occupying on the stack? since the default stack size is scarce by default (1 MB by MSVC).

**Comment by arvidn** [comment]
my immediate reaction is that the header normally is just a few hundred bytes. Of course any of those strings *could* be really large (`such as hostname`, `service_namespace`), but I don't think it's very important to support that. It's not clear to me whether this warning suggests an actual problem or not. Do you think there's a reason to believe those strings are *likely* to end up using 2047 bytes?

in other words, I consider the snprintf-truncation a security feature, to guard against maliciously large strings. On the other hand, since the GCC folks though it was a good idea to put this warning in, maybe they're on to something.

**Comment by Chocobo1** [comment]
>On the other hand, since the GCC folks though it was a good idea to put this warning in, maybe they're on to something.

Quoting their document:
>Level 1 of -Wformat-truncation enabled by -Wformat employs a conservative approach that warns only about calls to bounded functions whose return value is unused and that will most likely result in output truncation.

gcc probably employs some heuristics here, so it could be a false positive, also I noticed this warning is suppressed on mingw: https://github.com/arvidn/libtorrent/blob/ad540ead79c63bf04a5a05e675a950611a026ab6/include/libtorrent/config.hpp#L55

I cannot really say there will be truncation, but on the other hand if you believe this is false positive I think it is preferable to suppress the `Wformat-truncation` warning explicitly.
